### PR TITLE
Adds S3 storage backend

### PR DESCRIPTION
**Pull request by ArthurPBressan**
Adds a working backend implementation for AWS S3.

### CODE REVIEW

**Comment by mrjoes** [comment]
Can you put it in a separate file because of additional dependencies?

**Comment by pawl** [comment]
@ArthurPBressan can you also add an s3 backend example to the "Managing Files & Folders" section of the docs?: http://flask-admin.readthedocs.org/en/latest/advanced/#managing-files-folders

This is a really great idea, thanks for adding it. Might want to add your name to AUTHORS file too.

Next up, file admin pagination?

File: flask_admin/contrib/fileadmin/s3.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            path += self.separator
+        for key in self.bucket.list(path, self.separator):
+            if key.name == path:
+                continue
+            if isinstance(key, Prefix):
+                name = _remove_trailing_slash(_strip_path(key.name, path))
+                key_name = _remove_trailing_slash(key.name)
+                directories.append((name, key_name, True, 0, 0))
+            else:
+                last_modified = int(dateutil.parser.parse(key.last_modified)
```
Comment (pawl) [inline flask_admin/contrib/fileadmin/s3.py]: Can this be done with time.strptime and without the dateutil dependency?

File: flask_admin/contrib/fileadmin/s3.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+from flask_admin.babel import gettext
+
+
+class S3Storage(object):
+    """
+        Storage object representing files on an Amazon S3 bucket.
+
+        Usage::
+
+            from flask_admin.contrib.fileadmin import BaseFileAdmin,
```
Comment (pawl) [inline flask_admin/contrib/fileadmin/s3.py]: extra comma at the end of this line that can be removed

File: flask_admin/contrib/fileadmin/s3.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            from flask.admin.contrib.fileadmin.s3 import S3Storage
```
Comment (pawl) [inline flask_admin/contrib/fileadmin/s3.py]: `flask.admin` should be `flask_admin` like in the line above

**Comment by pawl** [comment]
I'm getting "Not Found" when I try to rename or delete files, but uploading works fine. Here's a gist of the code I'm using to test: https://gist.github.com/pawl/717fb9edde7bb6748427

**Comment by mrjoes** [comment]
Also, I'd suggest adding convenience class, like current FileAdmin - S3FileAdmin. This way users won't have to create storage instance and use BaseFileAdmin class.

**Comment by pawl** [comment]
@mrjoes :+1: was just about to say that. It would make usage even simpler.

File: flask_admin/contrib/fileadmin/s3.py (hunk 1 above)
Comment (pawl) [inline flask_admin/contrib/fileadmin/s3.py]: Here's an example that uses strptime: https://github.com/jordansissel/s3cleaner/blob/5969afaf5edf0223f20e13a21aa190d20c6b958a/s3cleaner.py#L46

**Comment by ArthurPBressan** [comment]
@pawl Should be working now. :+1:

**Comment by mrjoes** [comment]
@ArthurPBressan excellent, thank you.

Merged and will be in a new Flask-Admin release, which I was delaying for this feature.

**Comment by ArthurPBressan** [comment]
Glad to help. :smile:

**Pull request merged**